File: csharp/CsFind/CsFindLib/ArgTokenizer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace CsFindLib;

public class ArgTokenizer
{
    private Dictionary<string, string> BoolDictionary { get; } =  new();
    private Dictionary<string, string> StringDictionary { get; } = new();
    private Dictionary<string, string> IntDictionary { get; }  = new();
    private Dictionary<string, string> LongDictionary { get; }  = new();
    
    private static readonly Regex NumModifierPattern = new(@"^\d+[ckmgtp]$", RegexOptions.Compiled | RegexOptions.IgnoreCase);


    public ArgTokenizer(List<IOption> options)
    {
	    foreach (var option in options)
	    {
		    if (option.ArgType == ArgTokenType.Bool)
		    {
			    BoolDictionary[option.LongArg] = option.LongArg;
			    if (option.ShortArg != null)
			    {
				    BoolDictionary[option.ShortArg] = option.LongArg;
			    }
		    }
		    else if (option.ArgType == ArgTokenType.String)
		    {
			    StringDictionary[option.LongArg] = option.LongArg;
			    if (option.ShortArg != null)
			    {
				    StringDictionary[option.ShortArg] = option.LongArg;
			    }
		    }
		    else if (option.ArgType == ArgTokenType.Int) 
		    {
			    IntDictionary[option.LongArg] = option.LongArg;
			    if (option.ShortArg != null)
			    {
				    IntDictionary[option.ShortArg] = option.LongArg;
			    }
		    }
		    else if (option.ArgType == ArgTokenType.Long)
		    {
			    LongDictionary[option.LongArg] = option.LongArg;
			    if (option.ShortArg != null)
			    {
				    LongDictionary[option.ShortArg] = option.LongArg;
			    }
		    }
	    }
    }

    public ArgToken TokenizeSizeArg(string argName, string argVal)
    {
	    long lng;
	    try
	    {
		    lng = long.Parse(argVal);
	    }
	    catch (FormatException)
	    {
		    if (NumModifierPattern.IsMatch(argVal))
		    {
			    lng = argVal[..^1] switch
			    {
				    var s when int.TryParse(s, out var num) => num,
				    _ => throw new FindException($"Invalid value for option {argName}: {argVal}")
			    };
			    var modifier = argVal.ToLowerInvariant()[^1];
			    lng *= modifier switch
			    {
				    'c' => 1,
				    'k' => 1024,
				    'm' => 1024 * 1024,
				    'g' => 1024 * 1024 * 1024,
				    't' => 1024L * 1024 * 1024 * 1024,
				    'p' => 1024L * 1024 * 1024 * 1024 * 1024,
				    _ => throw new FindException($"Invalid value for option {argName}: {argVal}")
			    };
		    }
		    else
		    {
			    throw new FindException($"Invalid value for option {argName}: {argVal}");
		    }
	    }
	    return new ArgToken(argName, ArgTokenType.Long, lng);
    }

    public ArgToken TokenizeLongArg(string argName, string argVal)
    {
	    if (argName is "maxsize" or "minsize")
	    {
		    return TokenizeSizeArg(argName, argVal);
	    }
	    long lng;
	    try
	    {
		    lng = long.Parse(argVal);
	    }
	    catch (FormatException)
	    {
		    throw new FindException($"Invalid value for option {argName}: {argVal}");
	    }
	    return new ArgToken(argName, ArgTokenType.Long, lng);
    }
    
    public List<ArgToken> TokenizeArgs(IEnumerable<string> args)
	{
		var argTokens = new List<ArgToken>();
		var queue = new Queue<string>(args);

		while (queue.Count > 0)
		{
			var arg = queue.Dequeue();
			if (arg.StartsWith('-'))
			{
				var argNames = new List<string>();
				if (arg.StartsWith("--") && arg.Length > 2)
				{
					// Process long arg
					arg = arg[2..];
					// TODO: if = in arg, split
					if (arg.Contains('='))
					{
						var parts = arg.Split('=');
						if (parts.Length > 0 && !string.IsNullOrEmpty(parts[0]))
						{
							arg = parts[0];
						}
						if (parts.Length > 1 && !string.IsNullOrEmpty(parts[1]))
						{
							queue = new Queue<string>(queue.Prepend(parts[1]));
						}
					}
					argNames.Add(arg);
				}
				else if (arg.Length > 1)
				{
					// Process short arg(s)
					arg = arg[1..];
					foreach (var c in arg)
					{
						if (BoolDictionary.TryGetValue(c.ToString(), out var boolArg))
						{
							argNames.Add(boolArg);
						}
						else if (StringDictionary.TryGetValue(c.ToString(), out var stringArg))
						{
							argNames.Add(stringArg);
						}
						else if (IntDictionary.TryGetValue(c.ToString(), out var intArg))
						{
							argNames.Add(intArg);
						}
						else if (LongDictionary.TryGetValue(c.ToString(), out var longArg))
						{
							argNames.Add(longArg);
						}
						else
						{
							throw new FindException($"Invalid option: {c}");
						}
					}
				}

				foreach (var argName in argNames)
				{
					if (BoolDictionary.ContainsKey(argName))
					{
						argTokens.Add(new ArgToken(argName, ArgTokenType.Bool, true));
					}
					else if (StringDictionary.ContainsKey(argName)
					         || IntDictionary.ContainsKey(argName)
					         || LongDictionary.ContainsKey(argName)
					         || argName.Equals("settings-file"))
					{
						try
						{
							var argVal = queue.Dequeue();
							if (StringDictionary.ContainsKey(argName) || argName.Equals("settings-file"))
							{
								argTokens.Add(new ArgToken(argName, ArgTokenType.String, argVal));
							}
							else if (IntDictionary.ContainsKey(argName))
							{
								argTokens.Add(new ArgToken(argName, ArgTokenType.Int, int.Parse(argVal)));
							}
							else if (LongDictionary.ContainsKey(argName))
							{
								argTokens.Add(TokenizeLongArg(argName, argVal));
							}
							else
							{
								throw new FindException($"Invalid option: {arg}");
							}
						}
						catch (InvalidOperationException)
						{
							throw new FindException($"Missing value for option {arg}");
						}
					}
					else
					{
						throw new FindException($"Invalid option: {arg}");
					}
				}
			}
			else
			{
				argTokens.Add(new ArgToken("path", ArgTokenType.String, arg));
			}
		}
		return argTokens;
	}

	public List<ArgToken> TokenizeDictionary(Dictionary<string, object> dictionary)
	{
		var keys = dictionary.Keys.ToList();
		// keys are sorted so that output is consistent across all versions
		keys.Sort();
		var argTokens = new List<ArgToken>();
		foreach (var key in keys)
		{
			var val = dictionary[key];
			if (BoolDictionary.ContainsKey(key))
			{
				if (val is bool b)
				{
					argTokens.Add(new ArgToken(key, ArgTokenType.Bool, b));
				}
				else if (val is JsonElement { ValueKind: JsonValueKind.False or JsonValueKind.True } jsonBool)
				{
					argTokens.Add(new ArgToken(key, ArgTokenType.Bool, jsonBool.GetBoolean()));
				}
				else
				{
					throw new FindException($"Invalid value for option {key}: {val}");
				}
			}
			else if (StringDictionary.ContainsKey(key))
			{
				if (val is string s)
				{
					argTokens.Add(new ArgToken(key, ArgTokenType.String, s));
				}
				else if (val is IEnumerable<string> enumerable)
				{
					foreach (var enumStr in enumerable)
					{
						argTokens.Add(new ArgToken(key, ArgTokenType.String, enumStr));
					}
				}
				else if (val is JsonElement { ValueKind: JsonValueKind.String }  jsonElem)
				{
					var jsonStr = jsonElem.GetString();
					if (jsonStr is not null)
					{
						argTokens.Add(new ArgToken(key, ArgTokenType.String, jsonStr));
					}
					else
					{
						throw new FindException($"Invalid value for option {key}: {val}");
					}
				}
				else if (val is JsonElement { ValueKind: JsonValueKind.Array }  jsonArr)
				{
					foreach (var arrVal in jsonArr.EnumerateArray())
					{
						if (arrVal is JsonElement { ValueKind: JsonValueKind.String }  arrStr)
						{
							var jsonStr = arrStr.GetString();
							if (jsonStr is not null)
							{
								argTokens.Add(new ArgToken(key, ArgTokenType.String, jsonStr));
							}
							else
							{
								throw new FindException($"Invalid value for option {key}: {val}");
							}
						}
					}
				}
				else
				{
					throw new FindException($"Invalid value for option {key}: {val}");
				}
			}
			else if (IntDictionary.ContainsKey(key))
			{
				if (val is int i)
				{
					argTokens.Add(new ArgToken(key, ArgTokenType.Int, i));
				}
				else if (val is JsonElement { ValueKind: JsonValueKind.Number } jsonNum)
				{
					argTokens.Add(new ArgToken(key, ArgTokenType.Int, jsonNum.GetInt32()));
				}
				else
				{
					throw new FindException($"Invalid value for option {key}: {val}");
				}
			}
			else if (LongDictionary.ContainsKey(key))
			{
				if (val is long lng)
				{
					argTokens.Add(new ArgToken(key, ArgTokenType.Long, lng));
				}
				else if (val is JsonElement { ValueKind: JsonValueKind.Number } jsonNum)
				{
					argTokens.Add(new ArgToken(key, ArgTokenType.Long, jsonNum.GetInt64()));
				}
				else
				{
					throw new FindException($"Invalid value for option {key}: {val}");
				}
			}
			else
			{
				throw new FindException($"Invalid option: {key}");
			}
		}
		return argTokens;
	}

	public List<ArgToken> TokenizeJson(string jsonString)
	{
		try
		{
			var settingsDict = JsonSerializer.Deserialize<Dictionary<string, object>>(jsonString);
			if (settingsDict == null)
			{
				throw new FindException($"Unable to parse JSON");
			}
			return TokenizeDictionary(settingsDict);
		}
		catch (JsonException e)
		{
			throw new FindException($"Invalid JSON: {e.Message}");
		}
	}

	public List<ArgToken> TokenizeFile(string filePath)
	{
		var expandedPath = FileUtil.ExpandPath(filePath);
		var fileInfo = new FileInfo(expandedPath);
		if (!fileInfo.Exists)
			throw new FindException($"Settings file not found: {filePath}");
		if (!fileInfo.Extension.Equals(".json"))
			throw new FindException($"Invalid settings file (must be JSON): {filePath}");
		var contents = FileUtil.GetFileContents(expandedPath, Encoding.Default);
		return TokenizeJson(contents);
	}
}

File: fsharp/FsFind/FsFindLib/ArgTokenizer.fs
namespace FsFindLib

open System
open System.Collections.Generic
open System.IO
open System.Text
open System.Text.Json
open System.Text.RegularExpressions

type ArgTokenType = 
    | Bool = 0
    | Str = 1
    | Int = 2

type ArgToken = {
    Name : string;
    Type : ArgTokenType;
    Value : obj
}

type ArgTokenizer (boolMap: Map<string, string>, strMap: Map<string, string>, intMap: Map<string, string>) =

    let longArgWithValRegex = Regex("^--([a-zA-Z0-9-]+)=(.*)$")
    let longArgWithoutValRegex = Regex("^--([a-zA-Z0-9-]+)$")
    let shortArgsRegex = Regex("^-([a-zA-Z0-9]{2,})$")
    let shortArgRegex = Regex("^-([a-zA-Z0-9])$")

    let (|RegexMatch|_|) (regex:Regex) (arg:string) =            
        let m = regex.Match(arg)
        if m.Success then Some(m) else None

    member this.TokenizeArgs (args : string[]) : Result<ArgToken list, string> =
        let rec recTokenizeArgs (argList : string list) (argTokens : ArgToken list) : Result<ArgToken list, string> =
            match argList with
            | [] -> Ok argTokens
            | head :: tail ->
                match head with
                | RegexMatch longArgWithValRegex m ->
                    let longArg = m.Groups[1].Value
                    let argVal = m.Groups[2].Value
                    recTokenizeArgs (List.append ["--" + longArg; argVal] tail) argTokens
                | RegexMatch longArgWithoutValRegex m ->
                    let longArg = m.Groups[1].Value
                    if boolMap.ContainsKey(longArg) then
                        if longArg = "help" then
                            recTokenizeArgs [] [{ Name="help"; Type=ArgTokenType.Bool; Value=true }]
                        else
                            recTokenizeArgs tail (List.append argTokens [{ Name=longArg; Type=ArgTokenType.Bool; Value=true }])
                    elif strMap.ContainsKey(longArg) || intMap.ContainsKey(longArg) || longArg.Equals("settings-file") then
                        match tail with
                        | [] ->
                            Error $"Missing value for option: %s{longArg}"
                        | aHead :: aTail ->
                            if strMap.ContainsKey(longArg) then
                                recTokenizeArgs aTail (List.append argTokens [{ Name=longArg; Type=ArgTokenType.Str; Value=aHead }])
                            elif intMap.ContainsKey(longArg) then
                                recTokenizeArgs aTail (List.append argTokens [{ Name=longArg; Type=ArgTokenType.Int; Value=(int aHead) }])
                            else
                                match this.TokenizeFile aHead with
                                | Ok settingsFileTokens -> recTokenizeArgs aTail (List.append argTokens settingsFileTokens)
                                | Error e -> Error e
                    else
                        Error $"Invalid option: %s{longArg}"
                | RegexMatch shortArgsRegex m ->
                    let shortArgs = m.Groups[1].Value
                    let shortArgsList = shortArgs |> Seq.map (fun c -> "-" + string c) |> List.ofSeq
                    recTokenizeArgs (List.append shortArgsList tail) argTokens
                | RegexMatch shortArgRegex m ->
                    let shortArg = m.Groups[1].Value
                    if boolMap.ContainsKey(shortArg) then
                        let longArg = boolMap[shortArg]
                        recTokenizeArgs (List.append ["--" + longArg] tail) argTokens
                    elif strMap.ContainsKey(shortArg) then
                        let longArg = strMap[shortArg]
                        recTokenizeArgs (List.append ["--" + longArg] tail) argTokens
                    elif intMap.ContainsKey(shortArg) then
                        let longArg = intMap[shortArg]
                        recTokenizeArgs (List.append ["--" + longArg] tail) argTokens
                    else
                        Error $"Invalid option: %s{shortArg}"
                | _ ->
                    recTokenizeArgs tail (List.append argTokens [{ Name="path"; Type=ArgTokenType.Str; Value=head }])
        recTokenizeArgs (Array.toList args) []

    member this.TokenizeJsonElem (name : string) (jsonElem : JsonElement) : Result<ArgToken list, string> =
        if jsonElem.ValueKind = JsonValueKind.False || jsonElem.ValueKind = JsonValueKind.True then
            let b = jsonElem.GetBoolean()
            Ok [{ Name=name; Type=ArgTokenType.Bool; Value=b }]
        elif jsonElem.ValueKind = JsonValueKind.String then
            let s = jsonElem.GetString()
            if s = null then
                Error $"Invalid value for option: {name}"
            elif s = "settings-file" then
                this.TokenizeFile s
            else
                Ok [{ Name=name; Type=ArgTokenType.Str; Value=s }]
        elif jsonElem.ValueKind = JsonValueKind.Array then
            let rec recTokenizeJsonElems (elems : JsonElement list) (argTokens : ArgToken list) : Result<ArgToken list, string> =
                match elems with
                | [] -> Ok argTokens
                | elem :: tail ->
                    if elem.ValueKind = JsonValueKind.String then
                        let s = elem.GetString()
                        if s = null then
                            Error $"Invalid value for option: {name}"
                        // TODO: add settings-file handling here
                        else
                            recTokenizeJsonElems tail (List.append argTokens [{ Name=name; Type=ArgTokenType.Str; Value=s }])
                    else
                        Error $"Invalid value for option: {name}"
            let elems = jsonElem.EnumerateArray() |> List.ofSeq
            recTokenizeJsonElems elems []
        elif jsonElem.ValueKind = JsonValueKind.Number then
            let i = jsonElem.GetInt32()
            Ok [{ Name=name; Type=ArgTokenType.Int; Value=i }]
        else Error $"Invalid value for option: {name}"

    member this.TokenizeDictionary (dictionary : Dictionary<string, obj>) : Result<ArgToken list, string> =
        let rec recTokenizeDictionary (keys : string list) (dict : Dictionary<string, obj>) (argTokens : ArgToken list) : Result<ArgToken list, string> =
            match keys with
            | [] -> Ok argTokens
            | k :: tail ->
                match dict[k] with
                | :? JsonElement as jsonElem ->
                    match this.TokenizeJsonElem k jsonElem with
                    | Ok jsonElemTokens -> recTokenizeDictionary tail dict (List.append argTokens jsonElemTokens)
                    | Error e -> Error e
                | :? Boolean as b ->
                    recTokenizeDictionary tail dict (List.append argTokens [{ Name=k; Type=ArgTokenType.Bool; Value=b }])
                | :? string as s ->
                    if k = "settings-file" then
                        match this.TokenizeFile s with
                        | Ok settingsFileTokens -> recTokenizeDictionary tail dict (List.append argTokens settingsFileTokens)
                        | Error e -> Error e
                    else
                        recTokenizeDictionary tail dict (List.append argTokens [{ Name=k; Type=ArgTokenType.Str; Value=s }])
                | :? (string list) as strings ->
                    if k = "settings-file" then
                        match this.TokenizeFiles strings with
                        | Ok settingsFilesTokens -> recTokenizeDictionary tail dict (List.append argTokens settingsFilesTokens)
                        | Error e -> Error e
                    else
                        let newTokens = [ for s in strings do yield { Name=k; Type=ArgTokenType.Str; Value=s } ]
                        recTokenizeDictionary tail dict (List.append argTokens newTokens)
                | :? int as i ->
                    recTokenizeDictionary tail dict (List.append argTokens [{ Name=k; Type=ArgTokenType.Int; Value=i }])
                | _ -> Error $"Invalid value for option: {k}"

        // keys are sorted so that output is consistent across all versions
        let keys = dictionary.Keys |> List.ofSeq |> List.sort
        recTokenizeDictionary keys dictionary []

    member this.TokenizeJson (jsonString : string) : Result<ArgToken list, string> =
        match JsonSerializer.Deserialize<Dictionary<string, obj>>(jsonString) with
        | null -> Error "Unable to parse json"
        | settingsDict -> this.TokenizeDictionary settingsDict

    member this.TokenizeFile (filePath : string) : Result<ArgToken list, string> =
        let expandedPath = FileUtil.ExpandPath(filePath)
        let fileInfo = FileInfo(expandedPath)
        if fileInfo.Exists then
            if fileInfo.Extension.Equals(".json") then
                let contents = FileUtil.GetFileContents expandedPath Encoding.Default
                this.TokenizeJson contents
            else
                Error $"Invalid settings file (must be JSON): {filePath}"
        else
            Error $"Settings file not found: {filePath}"

    member this.TokenizeFiles (filePaths : string list) : Result<ArgToken list, string> =
        let rec recTokenizeFiles (filePaths : string list) (argTokens : ArgToken list) : Result<ArgToken list, string> =
            match filePaths with
            | [] -> Ok argTokens
            | fp :: tail ->
                match this.TokenizeFile fp with
                | Ok fileTokens -> recTokenizeFiles tail (List.append argTokens fileTokens)
                | Error e -> Error e
        recTokenizeFiles filePaths []

;;
